homeassistant/components/sleepiq/__init__.py

```diff
@@ -43,23 +54,48 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up the SleepIQ config entry."""
-    client = Sleepyq(entry.data[CONF_USERNAME], entry.data[CONF_PASSWORD])
+    conf = entry.data
+    email = conf[CONF_USERNAME]
+    password = conf[CONF_PASSWORD]
+
+    client_session = async_get_clientsession(hass)
+
+    gateway = AsyncSleepIQ(client_session=client_session)
+
     try:
-        await hass.async_add_executor_job(client.login)
-    except ValueError:
-        _LOGGER.error("SleepIQ login failed, double check your username and password")
+        await gateway.login(email, password)
+    except SleepIQLoginException:
+        _LOGGER.error("Could not authenticate with SleepIQ server")
         return False
+    except SleepIQTimeoutException as err:
+        raise ConfigEntryNotReady(
+            str(err) or "Timed out during authentication"
+        ) from err
 
-    coordinator = SleepIQDataUpdateCoordinator(
+    try:
+        await gateway.init_beds()
+    except SleepIQTimeoutException as err:
+        raise ConfigEntryNotReady(
+            str(err) or "Timed out during initialization"
+        ) from err
+    except SleepIQAPIException as err:
+        raise ConfigEntryNotReady(str(err) or "Error reading from SleepIQ API") from err
+
+    coordinator: DataUpdateCoordinator[None] = DataUpdateCoordinator(
         hass,
-        client=client,
-        username=entry.data[CONF_USERNAME],
+        _LOGGER,
+        name=f"SleepIQ Bed Statuses - {email}",
+        update_method=gateway.fetch_bed_statuses,
+        update_interval=UPDATE_INTERVAL,
     )
 
     # Call the SleepIQ API to refresh data
     await coordinator.async_config_entry_first_refresh()
 
-    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {
+        SLEEPIQ_DATA: gateway,
+        SLEEPIQ_STATUS_COORDINATOR: coordinator,
+    }
```

Comment:
Please keep the existing coordinator (preferred) or use a dataclass to store the data.